src/ring_theory/henselian.lean

```diff
@@ -0,0 +1,247 @@
+/-
+Copyright (c) 2021 Johan Commelin. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Johan Commelin
+-/
+
+import data.polynomial.taylor
+import ring_theory.ideal.local_ring
+import linear_algebra.adic_completion
+
+/-!
+# Henselian rings
+
+In this file we set up the basic theory of Henselian (local) rings.
+A ring `R` is *Henselian* at an ideal `I` if the following conditions hold:
+* `I` is contained in the Jacobson radical of `R`
+* for every polynomial `f` over `R`, with a *simple* root `a₀` over the quotient ring `R/I`,
+  there exists a lift `a : R` of `a₀` that is a root of `f`.
+  (Here, saying that a root `a` is *simple* means that `f.derivative.eval a` is a unit.)
+
+A local ring `R` is *Henselian* if it is Henselian at its maximal ideal.
+In this case the first condition is automatic, and in the second condition we may ask for
+`f.derivative.eval a ≠ 0`, since the quotient ring `R/I` is a field in this case.
+
+## Main declarations
+
+* `henselian_ring`: a typeclass on commutative rings,
+  asserting that the ring is Henselian at the ideal `I`.
+* `henselian_local_ring`: a typeclass on commutative rings,
+   asserting that the ring is local Henselian.
+* `field.henselian`: fields are Henselian local rings
+* `henselian.tfae`: equivalent ways of expressing the Henselian property for local rings
+* `is_adic_complete.henselian`:
+  a ring `R` with ideal `I` that is `I`-adically complete is Henselian at `I`
+
+## References
+
+https://stacks.math.columbia.edu/tag/04GE
+
+## Todo
+
+After a good API for etale ring homomorphisms has been developed,
+we can give more equivalent characterization os Henselian rings.
+
+In particular, this can give a proof that factorizations into coprime polynomials can be lifted
+from the residue field to the Henselian ring.
+
+The following gist contains some code sketches in that direction.
+https://gist.github.com/jcommelin/47d94e4af092641017a97f7f02bf9598
+
+-/
+
+
+
+noncomputable theory
+
+universe variables u v
+
+open_locale big_operators
+
+open local_ring polynomial function
+
+lemma is_local_ring_hom_of_le_jacobson_bot {R : Type*} [comm_ring R]
```

Comment:
An argument can be made that this should go in `ring_theory.jacobson_ideal`, but `is_local_ring_hom` isn't imported there. I don't mind either way, to be honest.

Reply:
I agree. At some point it will move to a proper home.